### PR TITLE
Use terrain singleton in WPNav and Location

### PR DESCRIPTION
**Pull request by peterbarker**
... instead of having the vehicles shove the terrain pointer in.

This is a significant change on Plane as it never used to set this pointer.  This might have affected terrain-relative missions in quadplanes while doing VTOL flight as Location would not be able to convert terrain-relative to home-relative-altitude, for example.

TODO: write an autotest for quadplanes doing terrain-relative VTOL flight

### CODE REVIEW

**Comment by rmackay9** [comment]
In generally this seems like a good thing to do.  I've certainly thought we should remove the _terrain pointer from AC_WPNav.  I'd prefer we don't do this until after s-curves goes in.. or make sure that we can easily rebase the scurve PR on top of this.

**Comment by peterbarker** [comment]
@rmackay9 the one conflict is trivially resolvable.

File: libraries/SITL/SIM_Aircraft.cpp [hunk 1]
```diff
@@ -94,15 +93,17 @@ void Aircraft::set_start_location(const Location &start_loc, const float start_y
...
 float Aircraft::ground_height_difference() const
 {
+#if AP_TERRAIN_AVAILABLE
+    AP_Terrain *terrain = AP::terrain();
     float h1, h2;
     if (sitl &&
-        sitl->terrain_enable && terrain &&
-        terrain->height_amsl(home, h1, false) &&
-        terrain->height_amsl(location, h2, false)) {
-        h2 += local_ground_level;
```
Comment (khancyr) [inline libraries/SITL/SIM_Aircraft.cpp:102]: whoups, you break something here!

**Review by khancyr — CHANGES_REQUESTED** [review]
there is some change on SITL that break the change of ground level calculation.
Otherwise it is good !

**Comment by khancyr** [comment]
This is a great change for dependencies ! it result is some great amount of flash save on each vehicle and surely on AP_Periph too !

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 2]
```diff
@@ -586,7 +587,9 @@ bool AC_WPNav::get_terrain_offset(float& offset_cm)
     case AC_WPNav::TerrainSource::TERRAIN_FROM_TERRAINDATABASE:
 #if AP_TERRAIN_AVAILABLE
         float terr_alt = 0.0f;
-        if (_terrain != nullptr && _terrain->height_above_terrain(terr_alt, true)) {
+        const AP_Terrain *terrain = AP::terrain();
```
Comment (khancyr) [inline libraries/AC_WPNav/AC_WPNav.cpp:590]: remove const

File: libraries/AP_Common/Location.cpp [hunk 3]
```diff
@@ -120,7 +118,11 @@ bool Location::get_alt_cm(AltFrame desired_frame, int32_t &ret_alt_cm) const
     float alt_terr_cm = 0;
     if (frame == AltFrame::ABOVE_TERRAIN || desired_frame == AltFrame::ABOVE_TERRAIN) {
 #if AP_TERRAIN_AVAILABLE
-        if (_terrain == nullptr || !_terrain->height_amsl(*(Location *)this, alt_terr_cm, true)) {
+        const AP_Terrain *terrain = AP::terrain();
```
Comment (khancyr) [inline libraries/AP_Common/Location.cpp:121]: remove const

File: libraries/SITL/SIM_Aircraft.cpp [hunk 4]
```diff
@@ -94,15 +93,20 @@ void Aircraft::set_start_location(const Location &start_loc, const float start_y
...
+        terrain != nullptr &&
+        sitl->terrain_enable &&
         terrain->height_amsl(home, h1, false) &&
         terrain->height_amsl(location, h2, false)) {
         h2 += local_ground_level;
         return h2 - h1;
     }
     return local_ground_level;
+#endif
+    return 0.0f;
```
Comment (khancyr) [inline libraries/SITL/SIM_Aircraft.cpp:109]: return local_ground_level;

Comment (khancyr) [reply libraries/SITL/SIM_Aircraft.cpp:109]: just move the return outside the #endif in fact

**Review by khancyr — CHANGES_REQUESTED** [review]
some small fix and it should be good !

**Review by khancyr — APPROVED** [review]
LGTM !

**Comment by rmackay9** [comment]
Yes, I think this looks OK..

**Comment by khancyr** [comment]
On contrary on what I said previously, it doesn't end to save more flash now, but that is still a good move in case we don't want terrain support.

File: libraries/AP_Common/Location.cpp [hunk 5]
```diff
@@ -120,7 +118,11 @@ bool Location::get_alt_cm(AltFrame desired_frame, int32_t &ret_alt_cm) const
     float alt_terr_cm = 0;
     if (frame == AltFrame::ABOVE_TERRAIN || desired_frame == AltFrame::ABOVE_TERRAIN) {
 #if AP_TERRAIN_AVAILABLE
-        if (_terrain == nullptr || !_terrain->height_amsl(*(Location *)this, alt_terr_cm, true)) {
+        AP_Terrain *terrain = AP::terrain();
+        if (terrain == nullptr) {
+            return false;
+        }
+        if (!terrain->height_amsl(*(Location *)this, alt_terr_cm, true)) {
```
Comment (tridge) [inline libraries/AP_Common/Location.cpp:125]: should be *this ?

File: libraries/SITL/SIM_Ship.cpp [hunk 6]
```diff
@@ -174,21 +174,15 @@ void ShipSim::send_report(void)
     Location loc = home;
     loc.offset(ship.position.x, ship.position.y);
 
-    int32_t alt;
-    bool have_alt = false;
+    int32_t alt = home.alt;  // assume home altitude
```
Comment (tridge) [inline libraries/SITL/SIM_Ship.cpp:177]: needs to be *10?

Comment (tridge) [reply libraries/SITL/SIM_Ship.cpp:177]: rename to alt_mm ?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Common/Location.cpp (hunk 5 above)
Comment (peterbarker) [reply libraries/AP_Common/Location.cpp:125]: Done in separate PR to see what happens.

File: libraries/SITL/SIM_Ship.cpp (hunk 6 above)
Comment (peterbarker) [reply libraries/SITL/SIM_Ship.cpp]: Done and done.

**Review by tridge — CHANGES_REQUESTED** [review]
we should default TERRAIN_ENABLE to 0 on sub

**Comment by peterbarker** [comment]
> we should default TERRAIN_ENABLE to 0 on sub

Done and tested in SITL

**Review by tridge — APPROVED** [review]

**Pull request merged**